File: Algorithms/C++/0002/0002.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode *head = new ListNode(0), *cur = head;
        while(l1 || l2) {
            int x = l1 ? l1->val : 0, y = l2 ? l2->val : 0;
            cur->next = new ListNode((x + y + carry) % 10); cur = cur->next;
            carry = (x + y + carry > 9) ? 1 : 0;
            l1 = l1 ? l1->next : NULL; l2 = l2 ? l2->next : NULL;
        }
        if(carry)
            cur->next = new ListNode(carry);
        return head->next;
    }
};

File: Algorithms/Python/0002/0002.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummyHead = ListNode();
        tail = dummyHead;
        carry = 0
        while l1 or l2:
            a = 0
            b = 0
            if l1:
                a = l1.val
                l1 = l1.next
            if l2:
                b = l2.val
                l2 = l2.next
            tail.next = ListNode((a + b + carry) % 10);
            tail = tail.next
            carry = 1 if a + b + carry >= 10 else 0
        tail.next = None if carry == 0 else ListNode(1)
        return dummyHead.next
